Email and flagging updates, upgrade to openstack sdk

JackPeisel opened the pull request:
Added a flagging feature wherein servers are first flagged(and if configured by the user, emails are sent out to the creators of such servers), and if the server isn't renamed by the next time cloud cleaner is called, then the server is deleted(if the force flag is set, as in earlier versions of cloud cleaner). Additionally shade and os-client-config have been phased out and upgraded to openstack sdk.  

NOTE: I plan to update the readme at a later point, I wanted to get a version finalized before touching changing it.

greg-hellings commented on cloud_cleaner/config.py:
```diff
@@ -128,14 +129,34 @@ def get_cloud(self):
+
+        :return: The dictionary
+        """
+        with open("email.json", "r") as reader:
```
What is `email.json` and why can't I configure that value?

greg-hellings commented on cloud_cleaner/resources/server.py:
```diff
@@ -111,6 +127,73 @@ def __right_name(self, target):
         system_age = system_age.replace(tzinfo=utc)
         return self._now > (system_age + self._interval)
+
+    def __set_flagged(self):
```
We had discussed the idea of keeping state, but I thought we opted to go with the simpler approach of just having the user call with a dates. Then they can do something like, "Dry-run for >2 weeks with email, real run for >3 weeks with email and deletion." That way we don't need to keep the state between runs.

JackPeisel replied:
I remember you mentioning that, I had changed it to require fewer calls(originally I had written it needing 2 calls of cloud-cleaner), but now I understand more what you're saying. I think that is a good idea, I was originally apprehensive because I was afraid of specific edge cases(using your example, cloud cleaner being called when a server is 1 week 6 days old, then being called when the same server is 3 weeks old, resulting in the server being deleted without a warning email ever sent), but thinking about it, I think most can be solved by just calling cloud cleaner regularly, which I believe the plan will be anyway. Also this change would reduce the number of additional files needed to use cloud-cleaner, which feels like a good thing.

greg-hellings commented on cloud_cleaner/resources/server.py:
```diff
@@ -111,6 +127,73 @@ def __right_name(self, target):
+        # List of server names that have been flagged to possibly be deleted
+        delete_list = []
+        # Open reader for the flagged file
+        reader = open(conn.name + "_flagged", 'r+')
```
If we go with state-keeping, then

1) This value should be configurable, so that the filename being written is not hard-coded
2) Use a context manager for opening files to avoid having unclosed files

greg-hellings commented on cloud_cleaner/resources/server.py:
```diff
@@ -111,6 +127,73 @@ def __right_name(self, target):
+        then sends emails if the flag is set in emails.json.
+        """
+        conn = self._get_conn()
+        writer = open(conn.name + "_flagged", 'w')
```
Same comment as the previous method regarding the naming and using a context manager.

greg-hellings commented on cloud_cleaner/resources/server.py:
```diff
@@ -111,6 +127,73 @@ def __right_name(self, target):
+        delete_list = []
+        # Open reader for the flagged file
+        reader = open(conn.name + "_flagged", 'r+')
+        for line in reader:
+            # delete the newline character at the end of the line
+            delete_list.append(line[:-1])
+
+        for id in delete_list:
+            server = conn.get_server(id)
+            self.__flagged.append(server)
```
You could write the meat of this method much more succinctly as:

```python
with open(conn.name + "_flagged", "r+") as reader:
    for line in reader:
        server = conn.get_server(line[:-1])
        self.__flagged.append(server)
    reader.write('')
```

greg-hellings commented on email.json:
```diff
@@ -0,0 +1,6 @@
+{
+  "Email": "Placeholder_for_Y/N",
+  "Sender": "Placeholder_for_str",
+  "smtp_server": "Placeholder_for_str",
+  "smtp_port": "Placeholder_for_int"
+}
```
I'm not sure what this file is doing, here. A little more info on its purpose, please?

JackPeisel replied:
email.json is meant to hold specific email configuration data, such as who should be the sender, what the smtp server should be, etc. I have the "Placeholder_for" to try to hint at what should be put there, for a user to then change. I was going to include a more in-depth explanation in an updated readme. That being said, if I'm making the change mentioned above, then I can also change from using a configuration file to using command line arguments for email configuration. I had originally done this, but eventually shied away from it thinking it was the cause of errors I was facing, and decided to keep my changes anyway when I realized the email arguments were not the root cause of the errors. I think I will change the email to using command line arguments instead, as it feels to me to go hand in hand with the above change I think is for the best about flagging/state.

greg-hellings commented on cloud_cleaner/config.py:
```diff
@@ -38,13 +37,28 @@ def __init__(self, parser=None, args=None):
                 "(up to 2 times)"
         self.__parser.add_argument("-v", "--verbose", help=_help,
                                    action='count', default=0)
+        _help = "Email warning messages to server creators if their server. " \
+                "is to be deleted. By default, the command does not email. " \
+                "Add this switch to include email."
+        self.__parser.add_argument("--email", "-e",
+                                   help=_help,
+                                   action='store_true')
+        _help = "The email address that should be used to send emails. Only " \
+                "used if --email is set. Required if --email is set."
+        self.__parser.add_argument("--sender", help=_help, default="")
+        _help = "The smtp server name which should be used to send emails. " \
+                "Only used if --email is set. Required if --email is set"
+        self.__parser.add_argument("--smtpN", help=_help, default="")
+        _help = "The smtp server port which should be used to send emails. " \
+                "Only used if --email is set. Required if --email is set"
+        self.__parser.add_argument("--smtpP", help=_help, default=0)
```
I wonder if it would look cleaner if we pushed these strings out to somewhere else. Something like:

```python
help_strings = {
    "email": "...",
    "sender": "...",
    ...
}
```

Then we could just do help=help_strings["email"] here. The "help_strings" variable could even be in a separate file that gets imported at the top `from help_strings import help_strings`?

Not a requirement, just an idea.

greg-hellings commented on cloud_cleaner/resources/fip.py:
```diff
@@ -36,10 +36,10 @@ def register(self, config):
 
-    def process(self):
-        shade = self._get_shade()
+    def process(self, deletion):
```
I don't know if I'm missing it, but I don't see "deletion" being used in this method anywhere.

greg-hellings commented on cloud_cleaner/resources/resource.py:
```diff
@@ -57,7 +57,7 @@ def register(self, config):
         self._sub_config = config.add_subparser(self.type_name)
 
-    def process(self):  # pylint: disable=no-self-use
+    def process(self, deletion):  # pylint: disable=no-self-use
```
Same as before - I don't see any usage of this variable.

greg-hellings commented on cloud_cleaner/resources/server.py:
```diff
@@ -36,16 +38,27 @@ def register(self, config):
-        self.__targets = shade.list_servers()
+        self.__age = self._config.get_arg('age')
+        self.__targets = conn.list_servers()
+        if(not deletion):
```
Parenthesis are out of place in Python when not needed for grouping or spacing across multiple lines.

greg-hellings commented on cloud_cleaner/resources/server.py:
```diff
@@ -54,6 +67,7 @@ def process(self):
         # Process for name
         self.__process_names()
         self._config.info("%d servers passed name test" % len(self.__targets))
+        self.__flagged = []
```
I don't think this is used, anymore, since we reduced it down to one call.

greg-hellings commented on cloud_cleaner/resources/server.py:
```diff
@@ -111,6 +133,34 @@ def __right_name(self, target):
+        for server in self.__targets:
+            user = conn.get_user_by_id(server.user_id, False)
+            # Cannot send an email to a user with no email
+            if(user.email is not None):
```
Unnecessary parentheses again.

greg-hellings commented on cloud_cleaner/resources/server.py:
```diff
@@ -111,6 +133,34 @@ def __right_name(self, target):
+            if(user.email is not None):
+                skip_name = self._config.get_arg("skip_name")
+                receiver = user.email
+                message = user.name + ", \n Your server, " + server.name
+                message = message + ''' may be deleted when its age reaches
+                    ''' + self.__age + ''' if you do not change the name of the
+                    server to include''' + skip_name + ''' at the start of the
+                    name. '''
```
Why not just make this a single block that you call .format() on?

greg-hellings commented:
Can we also add some tests to this code?

JackPeisel commented:
Fixed my use of parentheses, and updated some things from your (good) suggestions for making a single block to call .format() on, as well as the help strings. I changed the logic around slightly to get rid of the need for the "deletion" variable in places where it was unused. I updated the test cases, which I hadn't done earlier(updated them to reflect the update to openstack sdk from shade and etc, as well as some of the new logic of the process). There are not many tests to add, since the only major change that isn't covered by existed tests is the email/warning feature. I added a test to test this as best as I thought it could be(I check how many emails would have been sent out, and who they would have been sent out to). I will also be adding another test shortly which tests the email functionality simultaneously with the deletion of servers(with the force flag set).

greg-hellings commented on cloud_cleaner/bin/entrypoint.py:
```diff
@@ -26,9 +26,13 @@ def cloud_clean(args=sys.argv[1:],  # pylint: disable=W0102
         ALL_RESOURCES[config.get_resource()].clean()
     else:
         print("No changes made, force option not enabled")
+    ALL_RESOURCES[config.get_resource()].process()
```
Before I merge this, I would just like a little confirmation on this line: Is this going to be called twice (here and on L31) in the case of a call with `force == True`? It's not ideal, but I get it.

JackPeisel replied:
Yes it is going to be called twice in that case. I agree it isn't ideal, but I felt like it could still make sense. One process call before the actually deletion, so we know what to delete/clean(), and one after to know which servers to email or not(as well as to include a process call regardless of whether force == True, something in the current version of cloud cleaner).
I could move this process call to inside of the clean() function of server.py instead(at the end), which would keep functionality but could make the logic of entrypoint.py nicer if you would prefer that.
